pkg/provisioners/gcppubsub/util/logging/logging.go

```diff
@@ -0,0 +1,38 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// logging is a copy of knative/pkg's logging package, except it uses desugared loggers.
```

Comment:
I can probably contribute `With()`, as that isn't in knative/pkg right now and is useful. The other two methods are equivalent to existing knative/pkg functions, but with desugared logs. Do you think other people would want desugared logs?

Reply:
I prefer desugared.

I have accidentally misused the sugared logs multiple times and didn't notice until I was trying to debug and got annoyed that I didn't have the information that was needed.

I also find it useful to put the various things I want logged into named fields, rather than just inline in the message.

Reply:
Forgetting an argument when using the sugared logger is not caught by the compiler, instead causing a panic at runtime. The panic is recovered by zap and logged with an ugly error message, and the field value is lost.

```go
// right
logger.Infow("Failed to do the thing", "foo", "bar")
```

```json
{"level":"info","ts":1547845376.5430198,"logger":"fallback","caller":"containersource/reconcile.go:147","msg":"Failed to do the thing","foo":"bar"}
```

```go
// wrong
logger.Infow("Failed to do the thing", "foo")
```

```json
{"level":"dpanic","ts":1547845158.9617574,"logger":"fallback","caller":"zap/sugar.go:179","msg":"Ignored key without a value.","ignored":"foo","stacktrace":"github.com/knative/eventing-sources/pkg/controller/containersource.(*reconciler).setSinkURIArg\n\t/usr/local/google/home/grantrodgers/go/src/github.com/knative/eventing-sources/pkg/controller/containersource/reconcile.go:147\ngithub.com/knative/eventing-sources/pkg/controller/containersource.(*reconciler).Reconcile\n\t/usr/local/google/home/grantrodgers/go/src/github.com/knative/eventing-sources/pkg/controller/containersource/reconcile.go:84\ngithub.com/knative/eventing-sources/pkg/controller/testing.(*TestCase).Reconcile\n\t/usr/local/google/home/grantrodgers/go/src/github.com/knative/eventing-sources/pkg/controller/testing/table.go:156\ngithub.com/knative/eventing-sources/pkg/controller/testing.(*TestCase).Runner.func1\n\t/usr/local/google/home/grantrodgers/go/src/github.com/knative/eventing-sources/pkg/controller/testing/table.go:100\ntesting.tRunner\n\t/usr/lib/google-golang/src/testing/testing.go:827"}
{"level":"info","ts":1547845158.961737,"logger":"fallback","caller":"containersource/reconcile.go:147","msg":"Failed to do the thing"}
```



A similar error can happen when using `Infof`, though it doesn't cause a panic:

```go
logger.Infof("Failed to do the thing foo=%v")
```

```json
{"level":"info","ts":1547845576.9123294,"logger":"fallback","caller":"containersource/reconcile.go:147","msg":"Failed to do the thing %v"}
```

`fmt.Printf` cheats by having a special-case compiler check, but it doesn't work with other format strings:

```go
fmt.Printf("Failed to do the thing foo=%v")
```

```
./reconcile.go:149: Printf format %v reads arg #1, but call has 0 args
```

Out of all these, the desugared version is safest, fastest, and IMO most clear. Forgetting an argument or using the wrong type is a compile error, just like all normal function calls:

```go
logger := logger.Desugar()

// typed field
logger.Info("Failed to do the thing", zap.String("foo", "bar"))

// untyped field, equivalent to desugared or %v
logger.Info("Failed to do the thing", zap.Any("foo", "bar"))
```